### ISSUE TITLE
Database connection retry with back off pattern

### ISSUE THREAD

**Issue by bx80** [piwik/piwik#21807]
## Summary

On Matomo installations using scalable architecture database scaling events may sometimes cause short term connection unavailability for a few seconds which can cause queries to fail. 

For front end UI actions this is typically low impact as a page refresh will retry the query.
For the archiving process, a single query failure will cause the entire archive invalidation to be aborted and marked for later retry which will increase archiving time and resource usage.

Detecting temporary connection loss and simply retrying the query after a short delay can in many cases avoid these problems and increase the robustness of the application and archiving process.

## Proposed solution

In the PDO database driver catch 'server has gone away' exceptions and attempt to retry SELECT queries x times using a back off pattern where the delay before retrying increases for each attempt, after x unsuccessful attempts allow the exception to be thrown.

To provide full flexibility the following configuration options could be provided: 
```
; Retry this many times before giving up, set to zero to disable all retries
db_select_query_retry_attempts = 3

; Delay for these intervals before each retry attempt
db_select_query_retry_delays_seconds = [5,15,30]
```

**Comment by michalkleiner** [comment]
Nice, yeah, that can help and make the application more robust.
Would this only apply to SELECT queries or also INSERT or UPDATE? If they are run in transactions would they be automatically rolled back and it would be safe to try again?
If it did apply to other query types other than SELECT, we could adjust the config names, but I suspect that's intentional — correct?
For the latter I would add the word seconds to the comment to make it clear what unit it is.
Plus during the implementation I'd consider adding extra logging so that there is a trail of how many times, after how many retries, and in what time of day this happens most.

**Comment by bx80** [comment]
Good feedback @michalkleiner :+1:

I think this sort of automatic retry should just apply to SELECT queries, for INSERT, DELETE or UPDATE queries it would be better for the calling code to handle exceptions since there will be more context there.

I've updated the description to add the config unit to the name, and mention creating log messages.

Transactions could be a little tricky as we can't rollback the whole transaction and then retry a single select query as there have been other queries earlier in the transaction that would be missing. That said, as far as I can see we don't use transactions outside of tracking requests, so it might be acceptable to just check if a transaction is in progress  (eg. via `$connection->inTransaction();`) and only attempt automatic retries if not currently in a transaction.

**Comment by tsteur** [comment]
FYI on top of the transactions one thing that would need to be handled during archiving is the temporary segment table that would need to be re-created when this happens. These temp tables are bound to a connection so if a new connection is being created, the temp table would no longer exist.

**Comment by bx80** [comment]
That's a great point @tsteur :+1: This change might not be as straightforward as it first seemed, since many of the long running queries which have a higher chance of being interrupted will depend on temporary segment tables, then any worthwhile solution here will need to handle that too.

I don't think the DB adapter can or should know what temporary tables have been previously created, so it would be necessary to handle the retry in the code calling `LogAggregator::generateQuery()` (which itself calls the segment temporary table generation code). There are a limited number of places in core + common plugins that call `generateQuery()`, mostly in record builders and they generally follow a similar pattern, so we could add a new class to handle the retry logic and then wrap these code blocks with something like this:

```
$queryRetry = new QueryRetry();
do {
    try
    {
        // Existing code
        $query = $logAggregator::generateQuery( ... );

        // ... Do other stuff ...

        Db::getDb->query($query['sql'], $query['bind']);
        // end existing code
        
    } catch (Exception $e) {
        if ($queryRetry->doRetry($e) == true) {
            continue;
       }
       throw $e;
    }
    break;
} while (!$queryRetry->isDone());
```

QueryRetry class:
```
class QueryRetry() {

    private $retriesSetting;
    private $delaysSetting;
    private $attempts = 0;

    function __constructor() {
        // Get system settings
        $this->retriesSetting = [system setting from config];
        $this->delaysSetting = [system setting from config];
    }

    public function doRetry(Exception $e): bool {
        if ($e->message() == '.. server has gone away ..' && [other exceptions]
            && $this->attempts <= $this->retriesSetting
            && [other conditions]) 
        {
            $this->attempts++;
            sleep([with appropriate delay from $this->delaySetting]);
        }
        return false;
    }
 
    public function isDone() {
        if($this->attempts >= $this->retriesSetting) {
            return true;
        }
        return false;
    }
}
```
Does this general approach make sense?

**Comment by tsteur** [comment]
I don't have any thoughts on implementation. We just need to make sure that it works out of the box for all kind of scenarios and for plugins. Ideally, it would automatically work without any plugin developer needing to think about it or be aware of it. There may not be any harm if the DB adapter knew it had to recreate the temp table if it means it'll work in all cases and be easy and quicker to implement etc. The server gone away could happen during any kind of query like when fetching goals, etc. so all DB queries would maybe need to be rewritten which may not be worth the minimal impact.

On Cloud the issue happens very rarely (< 0.001% of the times). When it happens, it's in 85% of the cases during archiving. The other 15% is nothing to worry about. When this issue happens during archiving, it happens 99% of the time to the Writer Database connection when `insertArchiveData` is called. This means all the select queries on the reader finished and we want to insert data. This could indicate that the writer DB connection has been idle sitting around for a while waiting for the reads to finish. 

A fix for "Server gone away" may be a MySQL connection setting change or alternatively restarting the connection for the writer only. This simplifies it because the writer does not use temporary tables when a DB reader is configured. When no DB reader is configured, the issue likely doesn't happen because the same connection is constantly being used and not idle.
 
That means if we determine the issue is impactful to fix, we could only implement a reconnect logic while we're archiving, and also only when a DB reader is configured. Then we don't need to think about temporary tables, and I believe archiving also doesn't use transactions. The issue then becomes a lot more trivial to implement maybe.

Not sure it's clear what I mean. Something like below might be something to think about and since it might be fairly simple to implement it could be worth a try to see if it eliminates the issue (assuming we can't fix the issue with a MySQL config change).

```php
...
private $timeLastConnected = ...;
...
if (SettingsServer::isArchivePhpTriggered() && HasDbReaderConfigured && isRegularDbAndNotReaderDb && timeSinceLastConnect > 2 hours) {
  reconnectWriterDb
}
```

**Comment by bx80** [comment]
Since the cloud issue is occurring mostly in `insertArchiveData` then we should probably try the easier fix for that first, which if successful will reduce the priority of a general connection retry mechanism :+1:

If we do take a more general approach and we're okay having the DB adapter be aware of temporary tables, then a more seamless way of handling retries with temporary tables could be to have the segment table creation SQL cached in the DB adapter by the `GenerateQuery()` method and then cleared once the main query has completed, if the DB adapter needs to retry the connection then it could check if there is a cached segment table definition and re-execute it on the fresh connection before retrying the main query.

**Comment by tsteur** [comment]
> Since the cloud issue is occurring mostly in insertArchiveData then we should probably try the easier fix for that first, which if successful will reduce the priority of a general connection retry mechanism 👍

Indeed. As it seems to always happen in there, we can even just catch the error in that method and don't even need to change anything in the DB adapter. Minimises risk etc and a patch may be very simple.

**Comment by bx80** [comment]
@tsteur I've created a [new issue](https://github.com/matomo-org/matomo/issues/21873) to cover the quick solution :+1:
